pretixprint/app/src/main/java/eu/pretix/pretixprint/ui/MaintenanceFragment.kt

```diff
@@ -0,0 +1,193 @@
+package eu.pretix.pretixprint.ui
+
+import android.app.Dialog
+import android.os.Bundle
+import android.text.Editable
+import android.view.View.VISIBLE
+import android.view.inputmethod.EditorInfo
+import android.widget.ArrayAdapter
+import android.widget.AutoCompleteTextView
+import androidx.appcompat.app.AlertDialog
+import androidx.core.widget.addTextChangedListener
+import androidx.fragment.app.DialogFragment
+import androidx.lifecycle.lifecycleScope
+import androidx.preference.PreferenceManager
+import eu.pretix.pretixprint.R
+import eu.pretix.pretixprint.connections.*
+import eu.pretix.pretixprint.databinding.FragmentMaintenanceBinding
+import eu.pretix.pretixprint.util.DirectedHexdumpCollection
+import kotlinx.coroutines.*
+import java.io.DataInputStream
+import java.io.IOException
+
+private const val ARG_PRINTER_TYPE = "printer_type"
+
+class MaintenanceFragment : DialogFragment(R.layout.fragment_maintenance) {
+    enum class InputModes { HEX, ASCII }
+
+    private var printerType: String? = null
+    private var mode: InputModes = InputModes.HEX
+    private lateinit var binding: FragmentMaintenanceBinding
+    private lateinit var connection: ConnectionType
+    private var streamHolder: StreamHolder? = null
+    private var responseListener: Job? = null
+    private var hc = DirectedHexdumpCollection()
+
+    // FIXME: can the dialog close itself when app receives another intent?
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        arguments?.let {
+            printerType = it.getString(ARG_PRINTER_TYPE)
+        }
+        val prefs = PreferenceManager.getDefaultSharedPreferences(requireContext())
+        val con = prefs.getString("hardware_${printerType}printer_connection", "network_printer")
+        connection = getConnectionClass(con!!)!!
+    }
+
+    fun fail(message: String) {
+        binding.tvError.text = message
+        binding.tvError.visibility = VISIBLE
+        binding.btnSend.isEnabled = false
+    }
+
+    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
+        binding = FragmentMaintenanceBinding.inflate(layoutInflater)
+
+        val modes = InputModes.values().map { it.toString() }
+        val adapter = ArrayAdapter(requireContext(), android.R.layout.simple_dropdown_item_1line, modes)
+
+        val validateTilInput = fun(text: Editable?) {
+            if (mode == InputModes.HEX && text != null) {
+                if (text.replace(Regex("0[xX]"), "").contains(Regex("[^a-fA-F0-9 ]"))) {
+                    binding.tilInput.error = "Only hex characters allowed" // FIXME: extract string
+                    return
+                }
+                if (text.trim().replace(Regex("\\s+"), "").replace(Regex("0[xX]"), "").length % 2 != 0) {
```

Comment:
This means if I do

0x4 0x5

it will be interpreted as

45

instead of

0405

? Is that what a user would expect? No idea :D

(I'd be fine with not supporting the 0x syntax if that's the easiest fix)

Reply:
fixed in b4a9de4, requiring two characters after 0x helped :)